src/Board.cpp

```diff
@@ -0,0 +1,168 @@
+#include "Board.h"
+
+Board::Board(const std::string& board)
+	: m_playerIsWhite(true)
+{
+    Factory factory;
+    for (auto i = 0; i < board.size(); ++i)
+    {
+        auto row = i / ROWS;
+        auto col = i % COLS;
+        if (board[i] == 'K')
+            m_kingLoc[1] = { row, col };
+
+        else if (board[i] == 'k')
+            m_kingLoc[0] = { row, col };
+
+        m_board[row][col] = factory.createPiece(board[i]);
+    }
+}
+
+int Board::getMove(const std::string& move)
+{
+    Location src = { move[0] - 'a', move[1] - '1' };
+    Location dst = { move[2] - 'a', move[3] - '1' };
+
+    if (isEmpty(src))
+        return 11;
+    if (isRival(src))
+        return 12;
+    if (isCurrentPiecePlayer(dst))
+        return 13;
+    if (!m_board[src.x][src.y]->isCorrectMove(src, dst) || isWayBlocked(src, dst))
+        return 21;
+    if (moveCausesChess(src, dst))
+        return 31;
+    update(src, dst);
+    if (mate())
+        return 41;
+
+    return 42;
+}
+
+
+
+void Board::initBoard()
+{
+    for (size_t row = 0; row < ROWS; ++row)
+        for (size_t col = 0; col < COLS; ++col)
+            m_board[row][col] = nullptr;
+
+    m_kingLoc[0] = DEFULT_LOCATION;
+    m_kingLoc[1] = DEFULT_LOCATION;
+}
+
+bool Board::isEmpty(const Location source) const
+{
+    return m_board[source.x][source.y] == nullptr;
+}
+
+bool Board::isRival(const Location source) const
+{
+    return m_board[source.x][source.y]->getColor() != m_playerIsWhite;
+}
+
+bool Board::isCurrentPiecePlayer(const Location source) const
+{
+    return !isEmpty(source) && !isRival(source);
+}
+
+bool Board::isWayBlocked(const Location& source, const Location& destination) const
+{
+    int dx = destination.x - source.x;
+    int dy = destination.y - source.y;
+
+    if (dx != 0 && dy != 0 && std::abs(dx) != std::abs(dy))
+        return false;
+
+    int x_step = (dx > 0) ? 1 : ((dx < 0) ? -1 : 0);
+    int y_step = (dy > 0) ? 1 : ((dy < 0) ? -1 : 0);
+    Location currentLoc = source;
+    currentLoc.x += x_step;
+    currentLoc.y += y_step;
+    while (currentLoc.x != destination.x || currentLoc.y != destination.y) {
+        if (!isEmpty(currentLoc)) {
+            return true;
+        }
+        currentLoc.x += x_step;
+        currentLoc.y += y_step;
+    }
+    return false;
+
+}
+
+bool Board::moveCausesChess(const Location& source, const Location& destination)

```

Comment:
The function moveCausesChess should be marked as const since it doesn't modify the state of the Board object. This ensures that calling this function won't cause unexpected side effects. Try to think of a more appropriate name for the function